Add InvalidBounceTerrain parameter to Projectile

abc013 opened the pull request:
Closes #18742, but adds `InvalidBounceTerrain` instead of the requested `BounceTerrain`. This is because it makes more sense to me that one wants a projectile to bounce off all tiles but a specific one (in this case: water). However, if requested, I can also add `BounceTerrain`.

Testcase: Adds `InvalidBounceTerrain: Water` to the disk thrower in TS. However, it looks quite strange when the projectile explodes on coast tiles, which is why I wouldn't recommend using it for bleed.

abc013 committed:
Add InvalidBounceTerrain parameter to Projectile.

abcdefg30 reviewed (dismissed):
Lgtm.

Mailaender commented on OpenRA.Mods.Common/Projectiles/Bullet.cs:
```diff
@@ -93,6 +93,9 @@ public class BulletInfo : IProjectileInfo
 		public readonly string BounceSound = null;
 
+		[Desc("Terrain where the projectile explodes instead of bouncing.")]
+		public readonly string[] InvalidBounceTerrain = null;
```
```suggestion
		public readonly HashSet<string> InvalidBounceTerrain = new HashSet<string>();
```

seems to be the standard in the current code base. I guess because of performance.

abc013 replied:
Done.

abc013 committed:
Improve explosion determination.

abc013 commented:
While I was at it, I also removed the `shouldExplode` boolean and replaced it with `return` statments to prevent a tiny bit of unnecessary calculations (see second commit).

Mailaender commented:
You can remove https://github.com/OpenRA/OpenRA/pull/18779/commits/8d35a84e66b611866ca130f0e9f6e7b57431b401 now.

Mailaender commented on OpenRA.Mods.Common/Projectiles/Bullet.cs:
```diff
@@ -227,29 +235,37 @@ public void Tick(World world)
+				if (AnyValidTargetsInRadius(world, pos, info.Width, args.SourceActor, true))
+					return true;
+
 				target += (pos - source) * info.BounceRangeModifier / 100;
 				var dat = world.Map.DistanceAboveTerrain(target);
 				target += new WVec(0, 0, -dat.Length);
 				length = Math.Max((target - pos).Length / speed.Length, 1);
+
 				ticks = 0;
 				source = pos;
 				Game.Sound.Play(SoundType.World, info.BounceSound, source);
 				remainingBounces--;
```
The early return above change behavior here as these values won't be set?

phrohdoh commented:
Why would they remove the testcase before acquiring all the required approvals?

abcdefg30 commented:
Fwiw, the commit would still be accessible through the link in the comment.

abc013 commented:
I removed the testcase.

phrohdoh commented:
> Fwiw, the commit would still be accessible through the link in the comment.

Sure, but not nearly as easy to test (cannot just checkout the PR as-is now) and review.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9130ae92e9ca14c4b377a08b78deed8332289b91...97bd7f67259064bd3c0c08f151494770971a9973)
